crates/sui-core/src/execution_driver.rs

```diff
@@ -80,12 +82,25 @@ pub async fn execution_process(
         };
         authority.metrics.execution_driver_dispatch_queue.dec();
 
-        // TODO: Ideally execution_driver should own a copy of epoch store and recreate each epoch.
-        let epoch_store = authority.load_epoch_store_one_call_per_task();
-
         let digest = *certificate.digest();
         trace!(?digest, "Pending certificate execution activated.");
 
+        // A single execution driver process runs across epochs. After epoch change,
+        // transactions from the previous epoch cannot be executed correctly so it is better
+        // to skip execution gracefully.
+        // This should not happen on validators, but possibe on fullnodes with local execution
+        // in tests.
+        let epoch_store = authority.load_epoch_store_one_call_per_task();
+        if epoch != epoch_store.epoch() {
```

Comment:
Why is this check necessary?

Reply:
Not an expert in this code path, but IIUC only in local execution can authority advances to the next epoch where a transaction enqueued from a previous epoch has not executed. When that happens, before this change, execution would fail (validation failure) and panic after 10 retries. So it seems more graceful to skip execution instead.

Reply:
How could a transaction still be queued from a previous epoch and not executed, when we advance to the next epoch? Is that due to pruning?

Reply:
My understanding is that this is indeed impossible on validator, because validators wait for pending certificates to clear before advancing epoch. But on fullnode with local execution, I don't see anything preventing the following from happening:
1. Transaction orchestrator receives a quorum of signed effects, enqueue transaction for local execution.
2. Fullnode received end of epoch checkpoint via state sync, executes it and advances epoch.
3. The transaction enqueued at (1) gets into execution driver.

I may missed logic. Otherwise fullnode may need to add logic to wait for all local executions to finish before advancing epoch. Or since local execution will be removed, maybe we are ok with just a mitigation?

Reply:
But there are a few places where we explicitly check whether this cert has been executed in the execution process, and if so it returns Ok. So I am curious what was causing issues here. The only possible explanation would be due to aggressive pruning in tests, so that even if a transaction has been executed, it might think it hasn't.
btw this is not really related to this PR, right? Should we move it to a separate PR?